### ISSUE TITLE
End to end testing

### ISSUE THREAD

**Issue by Brett208** [OutpostUniverse/OP2BitmapConverter#12]
Provide a way to conduct end to end tests

- [x] Create sample tileset(s) using OP2Utility on disk for testing. Alternative could be to permanently commit sample files. This could be done for each end to end test or using the setup function call for gtest.
- Assist the test code in finding the path to the executable
    - [x] Visual Studio
    - [ ] makefile (blocked by issue #9)
- [x]  Use std::system to run the executable from test code and pass arguments. std::system does not support capturing stdout or stderr results. An alternative function could be the POSIX popen / _popen command which would support capturing the output but will have some platform specific issues to work through.

**Comment by DanRStevens** [comment]
We could use just one sample image file, perhaps a tall rectangle of about 2 game tiles worth, and a simple image or two in the cells, which are distinct from each other. We probably want at least one symbol that breaks vertical symmetry, and at least one symbol that breaks horizontal symmetry. Maybe just letters, like "A", and "B" (which are vertically and horizontally symmetric, but neither is both).

I think one image should be sufficient for success tests? Most tests will probably be for error cases where no image is generated.

Could use MS Paint to make a quick image. It doesn't need to be big or complicated.

As for path to executable, the test code could assume the executable is either in the current working directory, or on the path. The test runner code would then take care of setting up the environment (by either changing the current working directory, or setting the `PATH` environment variable) before running the test executable. The test harness should know what path to use based on the active system. For Linux and MacOS, that would be the `makefile`. For Windows, that could be the project settings. Though, are those settings part of the user specific files, which are normally excluded from version control?

I think I'd like to avoid copying executable files around. Takes time and disk space, and I don't think it would be required.

----

Another way to do unit tests is with some kind of script like a batch file, a shell file, or a makefile. I worked at a company that used makefiles to do end-to-end tests. Makefiles themselves are not platform specific, though the shell commands in them can be, depending on what they are. In practice, it was quite possible to write platform independent tests this way. The tests would be make goals, and consist of the command line to run the executable. Shell redirection can be used to capture `stdout` and `stderr`.

**Comment by Brett208** [comment]
The A / B bitmap sounds good.

If using a unit test project for Visual Studio builds, I don't think you can assume the non-test code executable is in the working directory or the path. Explained a bit more below. 

If we switch to some sort of batch file, the batch file could be initiated in the post build event which would know where to find the executable unlike a test code project. For the end to end tests, maybe the batch file makes the most sense. However I think Linux and Windows batch files are not compatible in syntax so we might need platform independent batch files? Wonder if there is an easy way around this... 

----- More on Visual Studio Executable Locations -----

On Visual Studio, the executable will be located something along the lines of:
/x64/Debug/OP2BitmapConverter.exe
/x64/Release/OP2BitmapConverter.exe
/Debug/OP2BitmapConverter.exe (Sometimes I will manually update the project file to prepend win32 or x86 to match the x64 layout, but have not yet for the BitmapConverter. The test project is setup this way)
/Release/OP2BitmapConverter.exe

The test project will be located at:
/test/x64/Debug/OP2BitmapConverterTest.exe
... etc ...

The post build event could use the `$(SolutionDir)` type syntax to locate the executable. I'm not sure how to find the executable in code without just setting all the hard coded values and then using combinations of macros like WIN32 and DEBUG to detect the right build environment (which isn't really a wrong solution, just brittle, although would be easy to correct if changed).

**Comment by DanRStevens** [comment]
That's true, the test executable would need to find the main executable. Though we might assume the test executable would be compiled with the same parameters as the main executable. Further, a project dependence would ensure the main executable is built before the test executable is built. That means we would have access to the `Configuration` and `Platform` variables in the test executable's environment, and they should correspond with an associated main executable.

I don't know that we need to worry about being platform independent with the test harness. The CI setup is unique per platform, so having a batch file for Windows testing, and a shell script or `makefile` target for Linux testing should be fine. I suppose technically a `makefile` is platform independent, just that it's not so common for someone to have `make` installed in a Windows environment.

**Comment by Brett208** [comment]
Test file added to master in PR #16

![TilesetTest](https://user-images.githubusercontent.com/15183337/107963412-22774d00-6f76-11eb-9ec6-507cd62a3253.png)

**Comment by Brett208** [comment]
End to end testing now exists in the master for Visual Studio builds. The Makefile portion is still pending. 

Noted in PR #16 the solution for finding the directory where Visual Studio places the executable could potentially be improved on in a future iteration.
